nixos/modules/services/networking/teleport.nix

```diff
@@ -0,0 +1,95 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.teleport;
+  settingsYaml = pkgs.formats.yaml { };
+  pidFilePath = "/run/teleport.pid";
+  cmdlineArgs = [ "--pid-file=${pidFilePath}" ]
+    ++ optional cfg.debug.enable "--debug"
+    ++ optional cfg.diag.enable "--diag-addr=${cfg.diag.addr}:${toString cfg.diag.port}"
+    ++ optional (cfg.settings != { }) "--config=${settingsYaml.generate "teleport.yaml" cfg.settings}"
+    ++ cfg.extraOpts;
+in
+{
+  options = {
+    services.teleport = with lib.types; {
+      enable = mkEnableOption "Teleport service";
+
+      settings = mkOption {
+        type = settingsYaml.type;
+        default = { };
+        example = literalExpression ''
+          {
+            teleport = {
+              nodename = "client";
+              advertise_ip = "192.168.1.2";
+              auth_token = "60bdc117-8ff4-478d-95e4-9914597847eb";
+              auth_servers = [ "192.168.1.1:3025" ];
+          }
+        '';
+        description = ''
+          Contents of the <literal>teleport.yaml</literal> config file.
+          The <literal>--config</literal> arguments will only be passed if this set is not empty.
+
+          See <link xlink:href="https://goteleport.com/docs/setup/reference/config/"/>.
+        '';
+      };
+
+      debug.enable = mkEnableOption "debug.";
+
+      diag = {
+        enable = mkEnableOption ''
+          endpoints for monitoring purposes. See <link xlink:href="https://goteleport.com/docs/setup/admin/troubleshooting/#troubleshooting/"/>
+        '';
+
+        addr = mkOption {
+          type = str;
+          default = "127.0.0.1";
+          description = "Metrics and diagnostics address.";
+        };
+
+        port = mkOption {
+          type = int;
+          default = 3000;
+          description = "Metrics and diagnostics port.";
+        };
+      };
+
+      extraOpts = mkOption {
+        type = listOf str;
+        default = [ ];
+        example = literalExpression ''
+          [
+            "--labels=Foo=Bar"
+          ]
+        '';
+        description = "Extra commandline options.";
+      };
+
+    };
+  };
+
+  config = mkIf config.services.teleport.enable {
+    environment.systemPackages = [ pkgs.teleport ];
+
+    systemd.services.teleport = {
```

Comment:
I'm not familiar with `teleport` at all so I apologize if this is a silly question, but I noticed the [upstream unit](https://github.com/gravitational/teleport/blob/740d184d1cfc69ae2e96c50ee738b13884fb232b/assets/monitoring/teleport.service) doesn't run as `root`. Is that something we can do?

Reply:
Hmm... or maybe [it does](https://github.com/gravitational/teleport/blob/740d184d1cfc69ae2e96c50ee738b13884fb232b/examples/systemd/fips/teleport.service)... :thinking:

Reply:
There are certain conditions when we `can` run as non-root, those are very limited use cases and imply some limitations (for example if we run `teleport` as non-root user we can't use `teleport/tsh` to ssh to this node - ssh needs `root`). Adding ability to switch user in the module will require a complexity of different assertions. Thant makes it more complicated and fragile.

Reply:
Understood. Thanks for the answer!